### ISSUE TITLE
throwError directive crashes arara

### ISSUE THREAD

**Issue by ArTourter** [cereda/arara#114]
I originally thought it was specific to the rule I am currently developing but trying to trigger an error in a directive result in a crash:

`% arara: lualatex: {interaction: 'blah'}`

for instance, instead of producing the error message 

> The provided interaction value is not valid.

instead, crashes arara with

```
I have spotted an error in rule 'lualatex' located at
'/usr/local/texlive/2018/texmf-dist/scripts/arara/rules'. I could
not evaluate the flag expression of one of the arguments. This
part is quite tricky, since it involves aspects of the underlying
expression language. I will do my best to help you in any way I
can. There are more details available on this exception:

DETAILS ---------------------------------------------------------
[Error: throwError('The provided interaction value is not
valid.'): The provided interaction value is not valid.]
[Near :
{... @{ ....}]
             ^
[Line: 1, Column: 1]

Total: 0.05 seconds

shell returned 2
```

**Comment by cereda** [comment]
`throwError` is an interesting method to say the least. :wink:

An error, for `arara`, is considered a critical event, so the tool halts immediately. This is by design. The tool also halts for execution failures, but in this case, the culprit is not the rule logic itself. Perhaps in the future (version 5.0 on) we can make `arara` be more forgiving on certain scenarios.

The error message is actually seen in the *details* section:

```
[Error: throwError('The provided interaction value is not
valid.'): The provided interaction value is not valid.]
```

 Admittedly, it is not too easy to spot the message, as it looks like a typical syntax error. :wink: Since the rule is evaluated at runtime, the underlying script framework reports the abnormal behaviour like this. It is not wrong *per se*, but I can see the inadequacy in displaying messages in the error section, at least from the user point of view.

That said, I think `throwError` is doing its job regardless of the presentation. :wink: There is a couple of things I can suggest in order to make a more forgiving experience regarding errors in the rule:

1. I still think that wrong parameter values should be reported, and the entire workflow should halt because of this abnormal scenario. In order to make the error clearer, a dialog box could be displayed before `throwError` is executed:

        showMessage(1, 'Houston, we have a problem!',
        'Oh no, the parameter has a wrong value!');
    
    Then, after the user acknowledges the message (the user has to click the button in order to make the dialog go away), the error is thrown in the terminal and the tool halts. For more details on UI methods, see section 6.7 of the manual, which covers the dialog boxes. :wink: 

2. A dialog could be used, as suggested in the previous item, but instead of throwing an error immediately, replace the problematic parameter value by a correct one, as a fallback. Then the execution will continue as intended, the tool won't halt, and the user will be notified about the wrong parameter value. :wink: 

Those are my suggestions, hope one of them makes sense. :wink: 

Cheerio!

**Comment by ArTourter** [comment]
Thanks for this Paulo,

I think the problem for me with the current behaviour is that it looks like the error is in the rule rather than in the user's call of the rule.

Since I would just be filtering the correct options for the qpdf command, what I might do for the rule I am writing is use `system.out.println` as a warning and pass the option to qpdf anyway which will then come back as failed. It is not optimal as the rule syntax is not the same as that of the program itself, but that's is the role of the `println` bit. I would rather not have a dependency on a windowing system.

I'll give that a try if that is an acceptable way to do things

Cheers

**Comment by cereda** [comment]
> I think the problem for me with the current behaviour is that it looks like the error is in the rule rather than in the user's call of the rule.

Yes, I agree with you. From a user perspective, it really looks out of place. However, as the rule itself can be seen as a script to be interpreted on runtime, an error semantically denotes an unwanted behaviour and that is why it is caught as such. I probably designed things while wearing my engineer hat. :wink: 

Future versions of `arara` could have better support for user-defined errors. I might go back to the drawing board and see what can be done on this regard. Personally, I think it's quite tricky to add another layer of error handling, but I always favour friendly user interaction. :wink: 

> Since I would just be filtering the correct options for the qpdf command, what I might do for the rule I am writing is use `system.out.println` as a warning and pass the option to qpdf anyway which will then come back as failed. It is not optimal as the rule syntax is not the same as that of the program itself, but that's is the role of the `println` bit. 

It works, I admit, but it breaks our beautiful console-based output. :rofl: My typographic OCD is already driving me bonkers just after reading this approach. The mad world we live in! I might need some hot cup of Lucida in order to make my heart not pound too strongly.

On a more serious note, one for an upcoming update. I might include a console-based messaging system set of methods. And as I plan this feature, I can include support for logging messages as well.

Just a humble comment: if `qpdf` fails anyway, I still think it's better to leave `arara` fail with `throwError` rather than issuing a system call. But hey, that's just me. :wink: 

> I would rather not have a dependency on a windowing system.

It is a bit intrusive, as all dialogs are modal (such that users must interact with the UI element before they can return to the parent application), but they do not add too much to the overall execution time (modulo fancy look and feels, but they are disabled by default) and look quite pleasant IMHO. :wink:
